### PR TITLE
Fix a typo in arpack.hpp + some issues in iso_c_binding

### PR DESCRIPTION
**Pull request by myroikon**
## Pull request purpose

Fix errors in order to use Arpack routines from C++ calls

## Detailed changes proposed in this pull request

- Fix typo in arpack.hpp. bmat for general matrix was "B" instead of "G"
- Fix in iso_c_binding (.F90 files) pass scalar arguments instead of array of chars which caused instant crashes

The above were tested in Windows using Microsoft Visual Studio 2017 & 2019 and Intel Fortran 2020 compiler.

### CODE REVIEW

File: SRC/icbacn.F90 [hunk 1]
```diff
@@ -7,7 +7,7 @@ subroutine cnaupd_c(ido, bmat, n, which, nev, tol, resid, ncv, v, ldv,&
   implicit none
 #include "arpackdef.h"
   integer(kind=c_int),                              intent(inout) :: ido
-  character(kind=c_char),       dimension(1),       intent(in)    :: bmat
+  character(kind=c_char),                           intent(in)    :: bmat
   integer(kind=c_int),          value,              intent(in)    :: n
```
Comment (fghoussen) [inline SRC/icbacn.F90]: Kill `dimension(1)` : sounds good if CI is OK ! Could you to that also for parpack (PARPACK/SRC/MPI/icbp*F90) ?

File: SRC/icbacn.F90 [hunk 2]
```diff
@@ -23,7 +23,15 @@ subroutine cnaupd_c(ido, bmat, n, which, nev, tol, resid, ncv, v, ldv,&
...
+  
+  character(len=2):: w
+  integer         :: i
+  
+  do i =1,2
+      w(i:i) = which(i)
+  end do
+
+  call cnaupd(ido, bmat, n, w, nev, tol, resid, ncv, v, ldv,&
               iparam, ipntr, workd, workl, lworkl, rwork, info)
```
Comment (fghoussen) [inline SRC/icbacn.F90]: Copy/convert array argument in local `character(len=2)` : sounds good too if CI is OK !
Could you to that also for parpack (PARPACK/SRC/MPI/icbp*F90) ?

File: SRC/icbacn.F90 (hunk 1 above)
Comment (myroikon) [reply SRC/icbacn.F90]: Ok! Done

File: SRC/icbacn.F90 (hunk 2 above)
Comment (myroikon) [reply SRC/icbacn.F90]: Ok! Done

**Comment by myroikon** [comment]
Are changes made ok in order for you to merge this pull request?

**Comment by sylvestre** [comment]
Yeah, looks great. Just two extra requests:

- would it be possible to have a test to make sure we don't regress?
- could you please document your change in the CHANGES files?

**Comment by fghoussen** [comment]
> would it be possible to have a test to make sure we don't regress?

Already heavily tested for arpack (`icb_arpack_c*`, mostly `arpackmm` : test many-many cases).
Lightly tested for parpack (`icb_parpack_c*`).

**Comment by sylvestre** [comment]
Yeah but I am a huge believer in regression tests.

**Comment by fghoussen** [comment]
> Yeah but I am a huge believer in regression tests.

You do well :D Would be good to have `parpackmm` (and `parpackSolver.hpp`) but this would need a templated **MPI** library : not sure such a library exists ! (likely tricky to use `petsc`)

**Comment by myroikon** [comment]
@sylvestre no problem doing the test you mean but I think I didn't quite catch what kind of test it is. I think I am a bit inexperienced in this field.

Sure I will add these changes in the file you mentioned.

**Comment by sylvestre** [comment]
Some examples:

* https://github.com/opencollab/arpack-ng/blob/master/TESTS/bug_1315_single.c
* https://github.com/opencollab/arpack-ng/blob/master/TESTS/bug_1315_double.c
* https://github.com/opencollab/arpack-ng/blob/master/TESTS/bug_1323.f

**Comment by fghoussen** [comment]
Note that this PR impacts only ICB : 

- `bug_1323.f` is pure fortran : no ICB here

- `bug_1315_single.c` and `bug 1315_double.c` : old-fashion hardcoded C-style way to link with F90 (= the "old" way you go if you have no ICB), or the cmake "FCMangle.h" trick (which wraps the same old-fashion-C-style way to do the job), no ICB here

-  `icb_arpack_c*` and `icb_parpack_c*` : same idea / test as `bug_1315_single.c` and `bug 1315_double.c`but now using ICB. There are only a few minimal cases tested here but with both arpack and **p**arpack.

- `arpackmm` (which includes `arpackSolver.hpp`) : heavy tests here all based on ICB. Here arpack only is tested, not **p**arpack (would need `petsc` ? or some other **MPI** linear solver that could handle both dense and sparse matrix - Eigen is used for `arpackmm` but has no MPI support)

Basically, all tests that are impacted by ICB are those which need to include `arpack.h[pp]` or `parpack.h[pp]`

```
>> git grep "arpack\.h*" *.c *.cpp *.hpp
EXAMPLES/MATRIX_MARKET/arpackSolver.hpp:#include "arpack.h"
EXAMPLES/MATRIX_MARKET/arpackmm.cpp:  // Solve with arpack.
PARPACK/TESTS/MPI/icb_parpack_c.c:#include "parpack.h"
PARPACK/TESTS/MPI/icb_parpack_cpp.cpp:#include "parpack.hpp"
TESTS/icb_arpack_c.c:#include "arpack.h"
TESTS/icb_arpack_cpp.cpp:#include "arpack.hpp"
```

Hope this helps

**Comment by fghoussen** [comment]
Quick look for arpack : 

- `icb_arpack_c` tests `d` and `z` flavors

- `icb_arpack_cpp` tests `s` and `d` flavors => replace `d` by `c` ? If so, all flavors will be covered

Quick look for parpack : 

- `icb_parpack_c` tests `d` and `z` flavors

- `icb_parpack_cpp` tests `s` and `c` flavors

All flavors seems covered here.

If you want to cover also `?s/?n` (sym/nonsym) flavors (as `arpackmm ` does), the number of cases almost double : would need twice more tests !... Maybe too overkilling (personal opinion) : replacing `d` by `c` in  `icb_arpack_cpp` test would already be a good move (if so use tol = 1 or you'll likely get accuracy problems with single precision)

**Comment by fghoussen** [comment]
Erratum : `icb_arpack_cpp` tests both `s` and `d`, and, `c` and `z` flavors. So all cases seem tested unless you want to add tests for sym/nonsym (`?s/?n`)

**Comment by myroikon** [comment]
> Erratum : `icb_arpack_cpp` tests both `s` and `d`, and, `c` and `z` flavors. So all cases seem tested unless you want to add tests for sym/nonsym (`?s/?n`)

I can easily perform tests based on this (icb_arpack_cpp) file. Basically I use calls through my program to dsaupd including "arpack.hpp" and it proves to be really solid. 

Is there any standard way to perform regression tests mentioned using `icb_arpack_cpp`? Could you give me some more details?

**Comment by fghoussen** [comment]
> Is there any standard way to perform regression tests

The tests must be relevant and end with relevant checks : you need to check for eigen values/vectors, so need to you _know_ them. In `icb_*arpack_c[pp]` you have a diagonal "1, 2, 3, 4, ..., N-1" matrix, so that, you know eigen values by construction (1, 2, 3, ..., N-1), and then, you can test them. In `arpackmm`, I use "real life" matrix: eigen values/vectors are _not_ known so a mat-mult "A.X - lamdba.X" operation is performed and checked (= 0) after calling arpack.

If you want to add regressions tests, you could : 

1. upgrade existing `icb` tests

2. add your own tests with your own checks (eigen values and/or vectors) at the end 

3. do both 1) and 2) ! :D 

If you want to go 2), then you could (__not__ tested : this may need to be reworked !) : 

- in `icb_arpack_c`, `ds` and `zn` flavors are already tested so you could add a `dn` flavor (AFASR, `zs` does not exist). You could go this way : copy ds, rename it dn, replace ds[ae]upd by d**n**[ae]upd. This _should_ work (not sure) as you ask for nonsym solve with a sym matrix : would look like this (not tested) :

```
>> git diff TESTS/icb_arpack_c.c
diff --git a/TESTS/icb_arpack_c.c b/TESTS/icb_arpack_c.c
index 613cf1f..511c81d 100644
--- a/TESTS/icb_arpack_c.c
+++ b/TESTS/icb_arpack_c.c
@@ -86,6 +86,66 @@ int ds() {
   return 0;
 }
 
+int dn() {
+  a_int ido = 0;
+  char bmat[] = "I";
+  a_int N = 1000;
+  char which[] = "LM";
+  a_int nev = 9;
+  double tol = 0;
+  double resid[N];
+  a_int ncv = 2 * nev + 1;
+  double V[ncv * N];
+  a_int ldv = N;
+  a_int iparam[11];
+  a_int ipntr[14];
+  double workd[3 * N];
+  a_int rvec = 1;
+  char howmny[] = "A";
+  double* d = (double*)malloc((nev + 1) * sizeof(double));
+  a_int select[ncv];
+  for (int i = 0; i < ncv; i++) select[i] = 1;
+  double z[(N + 1) * (nev + 1)];
+  a_int ldz = N + 1;
+  double sigma = 0;
+  int k;
+  for (k = 0; k < 3 * N; ++k) workd[k] = 0;
+  double workl[3 * (ncv * ncv) + 6 * ncv];
+  for (k = 0; k < 3 * (ncv * ncv) + 6 * ncv; ++k) workl[k] = 0;
+  a_int lworkl = 3 * (ncv * ncv) + 6 * ncv;
+  a_int info = 0;
+
+  iparam[0] = 1;
+  iparam[2] = 10 * N;
+  iparam[3] = 1;
+  iparam[4] = 0;  // number of ev found by arpack.
+  iparam[6] = 1;
+
+  while (ido != 99) {
+    /* call arpack like you would have, but, use dnaupd_c instead of dnaupd_ */
+    dnaupd_c(&ido, bmat, N, which, nev, tol, resid, ncv, V, ldv, iparam, ipntr,
+             workd, workl, lworkl, &info);
+
+    dMatVec(&(workd[ipntr[0] - 1]), &(workd[ipntr[1] - 1]));
+  }
+  if (iparam[4] != nev) return 1;  // check number of ev found by arpack.
+
+  /* call arpack like you would have, but, use dneupd_c instead of dneupd_ */
+  dneupd_c(rvec, howmny, select, d, z, ldz, sigma, bmat, N, which, nev, tol,
+           resid, ncv, V, ldv, iparam, ipntr, workd, workl, lworkl, &info);
+  int i;
+  for (i = 0; i < nev; ++i) {
+    printf("%f\n", d[i]);
+    /*eigen value order: smallest -> biggest*/
+    if (fabs(d[i] - (double)(1000 - (nev - 1) + i)) > 1e-6) {
+      free(d);
+      return 1;
+    }
+  }
+  free(d);
+  return 0;
+}
+
 void zMatVec(double _Complex* x, double _Complex* y) {
   int i;
   for (i = 0; i < 1000; ++i) y[i] = x[i] * (i + 1.0 + _Complex_I * (i + 1.0));
@@ -175,6 +235,11 @@ int main() {
 
   printf("------\n");
 
+  int rc = dn();  // arpack without debug.
+  if (rc != 0) return rc;
+
+  printf("------\n");
+
   // clang-format off
   debug_c(6, -6, 1,
           1, 1, 1, 1, 1, 1, 1,

``` 

- same idea with `icb_parack_c[pp]` : but replacing here pds[ae]up with pd**n**[ae]upd

```
>> git diff PARPACK/TESTS/MPI/icb_parpack_c.c
diff --git a/PARPACK/TESTS/MPI/icb_parpack_c.c b/PARPACK/TESTS/MPI/icb_parpack_c.c
index f73e3e0..948121a 100644
--- a/PARPACK/TESTS/MPI/icb_parpack_c.c
+++ b/PARPACK/TESTS/MPI/icb_parpack_c.c
@@ -92,6 +92,70 @@ int ds() {
   return 0;
 }
 
+int dn() {
+  a_int ido = 0;
+  char bmat[] = "I";
+  a_int N = 1000;
+  char which[] = "LM";
+  a_int nev = 3;
+  double tol = 0;
+  double resid[N];
+  a_int ncv = 2 * nev + 1;
+  double V[ncv * N];
+  a_int ldv = N;
+  a_int iparam[11];
+  a_int ipntr[14];
+  double workd[3 * N];
+  a_int rvec = 1;
+  char howmny[] = "A";
+  double* d = (double*)malloc((nev + 1) * sizeof(double));
+  a_int select[ncv];
+  for (int i = 0; i < ncv; i++) select[i] = 1;
+  double z[(N + 1) * (nev + 1)];
+  a_int ldz = N + 1;
+  double sigma = 0;
+  int k;
+  for (k = 0; k < 3 * N; ++k) workd[k] = 0;
+  double workl[3 * (ncv * ncv) + 6 * ncv];
+  for (k = 0; k < 3 * (ncv * ncv) + 6 * ncv; ++k) workl[k] = 0;
+  a_int lworkl = 3 * (ncv * ncv) + 6 * ncv;
+  a_int info = 0;
+  int rank;
+  MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+
+  iparam[0] = 1;
+  iparam[2] = 10 * N;
+  iparam[3] = 1;
+  iparam[4] = 0;  // number of ev found by arpack.
+  iparam[6] = 1;
+
+  MPI_Fint MCW = MPI_Comm_c2f(MPI_COMM_WORLD);
+  while (ido != 99) {
+    /* call arpack like you would have, but, use dnaupd_c instead of dnaupd_ */
+    pdnaupd_c(MCW, &ido, bmat, N, which, nev, tol, resid, ncv, V, ldv, iparam,
+              ipntr, workd, workl, lworkl, &info);
+
+    dMatVec(&(workd[ipntr[0] - 1]), &(workd[ipntr[1] - 1]));
+  }
+  if (iparam[4] != nev) return 1;  // check number of ev found by arpack.
+
+  /* call arpack like you would have, but, use dneupd_c instead of dneupd_ */
+  pdneupd_c(MCW, rvec, howmny, select, d, z, ldz, sigma, bmat, N, which, nev,
+            tol, resid, ncv, V, ldv, iparam, ipntr, workd, workl, lworkl,
+            &info);
+  int i;
+  for (i = 0; i < nev; ++i) {
+    printf("rank %d - %f\n", rank, d[i]);
+    /*eigen value order: smallest -> biggest*/
+    if (fabs(d[i] - (double)(1000 - (nev - 1) + i)) > 1e-6) {
+      free(d);
+      return 1;
+    }
+  }
+  free(d);
+  return 0;
+}
+
 void zMatVec(double _Complex* x, double _Complex* y) {
   int i;
   for (i = 0; i < 1000; ++i) y[i] = x[i] * (i + 1.0 + _Complex_I * (i + 1.0));
@@ -190,6 +254,12 @@ int main() {
   MPI_Comm_rank(MPI_COMM_WORLD, &rank);
   if (rank == 0) printf("------\n");
 
+  int rc = dn();  // parpack without debug.
+  fflush(stdout);
+  MPI_Barrier(MPI_COMM_WORLD);
+  if (rc != 0) return rc;
+  if (rank == 0) printf("------\n");
+
   // clang-format off
   debug_c(6, -6, 1,
           1, 1, 1, 1, 1, 1, 1,

```

- you could also templatize `icb_parpack_cpp` (`icb_arpack_cpp` is already templatized). Doing would test **only** sym cases, but, you can tests **all** sym flavors (`ss`, `ds`, `cs`, `zs` - but no `?n`). Would look like (not tested)

```
diff --git a/PARPACK/TESTS/MPI/icb_parpack_cpp.cpp b/PARPACK/TESTS/MPI/icb_parpack_cpp.cpp
index ce5f367..eb5bcbd 100644
--- a/PARPACK/TESTS/MPI/icb_parpack_cpp.cpp
+++ b/PARPACK/TESTS/MPI/icb_parpack_cpp.cpp
@@ -26,6 +26,7 @@ void diagonal_matrix_vector_product(float const* const x, float* const y) {
   }
 }
 
+template <typename Real>
 void real_symmetric_runner() {
   a_int N = 1000;
   a_int nev = 3;
@@ -35,17 +36,17 @@ void real_symmetric_runner() {
   a_int ldv = N;
 
   a_int rvec = 1;
-  float tol = 0.0f;
-  float sigma = 0.0f;
+  Real tol = 0.0f;
+  Real sigma = 0.0f;
 
   std::array<a_int, 14> ipntr;
 
-  std::vector<float> workd(3 * N, 0.0f);
-  std::vector<float> workl(3 * (ncv * ncv) + 6 * ncv, 0.0f);
-  std::vector<float> V(ncv * N);
-  std::vector<float> d(nev + 1);
-  std::vector<float> z((N + 1) * (nev + 1));
-  std::vector<float> resid(N);
+  std::vector<Real> workd(3 * N, 0.0f);
+  std::vector<Real> workl(3 * (ncv * ncv) + 6 * ncv, 0.0f);
+  std::vector<Real> V(ncv * N);
+  std::vector<Real> d(nev + 1);
+  std::vector<Real> z((N + 1) * (nev + 1));
+  std::vector<Real> resid(N);
   std::vector<a_int> select(ncv);
   for (int i = 0; i < ncv; i++) select[i] = 1;
 
@@ -91,7 +92,7 @@ void real_symmetric_runner() {
   for (int i = 0; i < nev; ++i) {
     std::cout << "rank " << rank << " - " << d[i] << std::endl;
     /*eigen value order: smallest -> biggest*/
-    if (std::abs(d[i] - static_cast<float>(1000 - (nev - 1) + i)) > 1.) {
+    if (std::abs(d[i] - static_cast<Real>(1000 - (nev - 1) + i)) > 1.) {
       throw std::domain_error("Correct eigenvalues not computed");
     }
   }
@@ -104,6 +105,7 @@ void diagonal_matrix_vector_product(std::complex<float> const* const x,
   }
 }
 
+template <typename Real>
 void complex_symmetric_runner() {
   a_int N = 1000;
   a_int nev = 1;
@@ -111,23 +113,23 @@ void complex_symmetric_runner() {
   a_int ldv = N;
   a_int ldz = N + 1;
 
-  float tol = 0.0f;
+  Real tol = 0.0f;
   a_int rvec = 0;
-  std::complex<float> sigma(0.0f, 0.0f);
+  std::complex<Real> sigma(0.0f, 0.0f);
 
-  std::vector<std::complex<float>> resid(N);
-  std::vector<std::complex<float>> V(ncv * N);
-  std::vector<std::complex<float>> workd(3 * N);
-  std::vector<std::complex<float>> d(nev + 1);
-  std::vector<std::complex<float>> z((N + 1) * (nev + 1));
+  std::vector<std::complex<Real>> resid(N);
+  std::vector<std::complex<Real>> V(ncv * N);
+  std::vector<std::complex<Real>> workd(3 * N);
+  std::vector<std::complex<Real>> d(nev + 1);
+  std::vector<std::complex<Real>> z((N + 1) * (nev + 1));
   std::vector<a_int> select(ncv);
   for (int i = 0; i < ncv; i++) select[i] = 1;
 
   a_int lworkl = 3 * (ncv * ncv) + 6 * ncv;
-  std::vector<std::complex<float>> workl(lworkl);
+  std::vector<std::complex<Real>> workl(lworkl);
 
-  std::vector<std::complex<float>> rwork(ncv);
-  std::vector<std::complex<float>> workev(2 * ncv);
+  std::vector<std::complex<Real>> rwork(ncv);
+  std::vector<std::complex<Real>> workev(2 * ncv);
 
   a_int info = 0;
 
@@ -176,9 +178,9 @@ void complex_symmetric_runner() {
     std::cout << "rank " << rank << " - " << std::real(d[i]) << " "
               << std::imag(d[i]) << '\n';
     /*eigen value order: smallest -> biggest*/
-    if (std::abs(std::real(d[i]) - static_cast<float>(1000 - (nev - 1) + i)) >
+    if (std::abs(std::real(d[i]) - static_cast<Real>(1000 - (nev - 1) + i)) >
             1. ||
-        std::abs(std::imag(d[i]) - static_cast<float>(1000 - (nev - 1) + i)) >
+        std::abs(std::imag(d[i]) - static_cast<Real>(1000 - (nev - 1) + i)) >
             1.) {
       throw std::domain_error("Correct eigenvalues not computed");
     }
@@ -192,7 +194,17 @@ int main() {
 
   try {
     // parpack without debug
-    real_symmetric_runner();
+    real_symmetric_runner<float>();
+  } catch (std::domain_error& e) {
+    std::cout << e.what() << '\n';
+    MPI_Abort(MPI_COMM_WORLD, 1);
+  }
+
+  MPI_Barrier(MPI_COMM_WORLD);
+
+  try {
+    // parpack without debug
+    real_symmetric_runner<double>();
   } catch (std::domain_error& e) {
     std::cout << e.what() << '\n';
     MPI_Abort(MPI_COMM_WORLD, 1);
@@ -224,7 +236,16 @@ int main() {
           1);
 
   try {
-    complex_symmetric_runner();
+    complex_symmetric_runner<float>();
+  } catch (std::domain_error& e) {
+    std::cout << e.what() << '\n';
+    MPI_Abort(MPI_COMM_WORLD, 1);
+  }
+
+  MPI_Barrier(MPI_COMM_WORLD);
+
+  try {
+    complex_symmetric_runner<double>();
   } catch (std::domain_error& e) {
     std::cout << e.what() << '\n';
     MPI_Abort(MPI_COMM_WORLD, 1);

```

All this is not tested, and, will likely need some rework. Hope this helps

**Comment by fghoussen** [comment]
Also for `icb_parpack_cpp` you'll likely need a custom `tol_check` (single precision will likely need eps = 1, double precision should allow up to eps = 1e-06)

```
diff --git a/PARPACK/TESTS/MPI/icb_parpack_cpp.cpp b/PARPACK/TESTS/MPI/icb_parpack_cpp.cpp
index eb5bcbd..57bc462 100644
--- a/PARPACK/TESTS/MPI/icb_parpack_cpp.cpp
+++ b/PARPACK/TESTS/MPI/icb_parpack_cpp.cpp
@@ -27,7 +27,7 @@ void diagonal_matrix_vector_product(float const* const x, float* const y) {
 }
 
 template <typename Real>
-void real_symmetric_runner() {
+void real_symmetric_runner(double const & tol_check) {
   a_int N = 1000;
   a_int nev = 3;
   a_int ncv = 2 * nev + 1;
@@ -92,7 +92,7 @@ void real_symmetric_runner() {
   for (int i = 0; i < nev; ++i) {
     std::cout << "rank " << rank << " - " << d[i] << std::endl;
     /*eigen value order: smallest -> biggest*/
-    if (std::abs(d[i] - static_cast<Real>(1000 - (nev - 1) + i)) > 1.) {
+    if (std::abs(d[i] - static_cast<Real>(1000 - (nev - 1) + i)) > tol_check) {
       throw std::domain_error("Correct eigenvalues not computed");
     }
   }
@@ -106,7 +106,7 @@ void diagonal_matrix_vector_product(std::complex<float> const* const x,
 }
 
 template <typename Real>
-void complex_symmetric_runner() {
+void complex_symmetric_runner(double const & tol_check) {
   a_int N = 1000;
   a_int nev = 1;
   a_int ncv = 2 * nev + 1;
@@ -179,9 +179,9 @@ void complex_symmetric_runner() {
               << std::imag(d[i]) << '\n';
     /*eigen value order: smallest -> biggest*/
     if (std::abs(std::real(d[i]) - static_cast<Real>(1000 - (nev - 1) + i)) >
-            1. ||
+            tol_check ||
         std::abs(std::imag(d[i]) - static_cast<Real>(1000 - (nev - 1) + i)) >
-            1.) {
+            tol_check) {
       throw std::domain_error("Correct eigenvalues not computed");
     }
   }
@@ -194,7 +194,7 @@ int main() {
 
   try {
     // parpack without debug
-    real_symmetric_runner<float>();
+    real_symmetric_runner<float>(1);
   } catch (std::domain_error& e) {
     std::cout << e.what() << '\n';
     MPI_Abort(MPI_COMM_WORLD, 1);
@@ -204,7 +204,7 @@ int main() {
 
   try {
     // parpack without debug
-    real_symmetric_runner<double>();
+    real_symmetric_runner<double>(1.-06);
   } catch (std::domain_error& e) {
     std::cout << e.what() << '\n';
     MPI_Abort(MPI_COMM_WORLD, 1);
@@ -236,7 +236,7 @@ int main() {
           1);
 
   try {
-    complex_symmetric_runner<float>();
+    complex_symmetric_runner<float>(1.);
   } catch (std::domain_error& e) {
     std::cout << e.what() << '\n';
     MPI_Abort(MPI_COMM_WORLD, 1);
@@ -245,7 +245,7 @@ int main() {
   MPI_Barrier(MPI_COMM_WORLD);
 
   try {
-    complex_symmetric_runner<double>();
+    complex_symmetric_runner<double>(1.e-06);
   } catch (std::domain_error& e) {
     std::cout << e.what() << '\n';
     MPI_Abort(MPI_COMM_WORLD, 1);


```

**Comment by myroikon** [comment]
@fghoussen thank you for your help and sorry for the belated response. I will update the tests in icb_*arpack_c[pp] files and let you know about the results. 

Also in our application, we use extensively the solver for the symmetric cases after the changes I made and the results are verified with results from an independent solver we use.

**Comment by fghoussen** [comment]
> I will update the tests in icb_*arpack_c[pp] files

Thanks ! Prefer pushing step-by-step (!) : likely have to deal with / fix problems along the way...

> Also in our application, we use extensively the solver for the symmetric cases after the changes I made and the results are verified with results from an independent solver we use.

BTW, if you use `pyarpack` (based on `arpackSolver.hpp`) and/or `arpackSolver.hpp`, and, that you have the opportunity to compare with other solvers (+ checking with same accuracy !), do not hesitate to report about timings. At first, I wrote `arpackSolver.hpp` and `arpackmm` for "testing around" (playing with parameters : shift, solvers, ...), but, not for performance : having a (small, or beginning of) benchmark could be a good thing in order to maybe improve performance in the future !...

**Comment by myroikon** [comment]
> BTW, if you use pyarpack (based on arpackSolver.hpp) and/or arpackSolver.hpp

pyarpack is not used but arpackSolver.hpp is included in our C++ project through which arpack::saupd and arpack::seupd routines are called. 

> and, that you have the opportunity to compare with other solvers (+ checking with same accuracy !)

our in-house solver is used, which is based on Jacobi method, to verify the results but in terms of performance, it is not the best so ARPACK by far outperforms it.

**Comment by fghoussen** [comment]
> so ARPACK by far outperforms it.

OK ! :D

**Comment by sylvestre** [comment]
@fghoussen do you think we should merge that ? (besides the conflicts)

**Comment by myroikon** [comment]
> @fghoussen do you think we should merge that ? (besides the conflicts)

Let me update my branch and resolve the conflicts. I was a little bit busy the last two months so I wasn't able to update the examples - tests. If you think so I could try to update them the following week.

**Comment by sylvestre** [comment]
Does it have an impact on the ABI ?
ie, will user of arpack will have to rebuild their program with this version?
thanks

**Comment by fghoussen** [comment]
No https://github.com/opencollab/arpack-ng/pull/155#issuecomment-422372619

**Comment by sylvestre** [comment]
Merci!
